runtime-sdk/src/modules/consensus_accounts/mod.rs

```diff
@@ -450,34 +461,77 @@ impl<Accounts: modules::accounts::API, Consensus: modules::consensus::API>
     fn tx_delegate<C: TxContext>(ctx: &mut C, body: types::Delegate) -> Result<(), Error> {
         let params = Self::params();
         <C::Runtime as Runtime>::Core::use_tx_gas(ctx, params.gas_costs.tx_delegate)?;
+        let store_receipt = body.receipt > 0;
+        if store_receipt {
+            <C::Runtime as Runtime>::Core::use_tx_gas(ctx, params.gas_costs.store_receipt)?;
+        }
 
         // Check whether delegate is allowed.
         if params.disable_delegate {
             return Err(Error::Forbidden);
         }
+        // Make sure receipts can only be requested internally (e.g. via subcalls).
+        if store_receipt && !ctx.is_internal() {
+            return Err(Error::InvalidArgument);
+        }
 
         // Signer.
         let signer = &ctx.tx_auth_info().signer_info[0];
         let from = signer.address_spec.address();
-        let nonce = signer.nonce;
-        Self::delegate(ctx, from, nonce, body.to, body.amount)
+        let nonce = if store_receipt {
+            body.receipt // Use receipt identifier as the nonce.
+        } else {
+            signer.nonce // Use signer nonce as the nonce.
+        };
+        Self::delegate(ctx, from, nonce, body.to, body.amount, store_receipt)
     }
 
     #[handler(call = "consensus.Undelegate")]
     fn tx_undelegate<C: TxContext>(ctx: &mut C, body: types::Undelegate) -> Result<(), Error> {
         let params = Self::params();
         <C::Runtime as Runtime>::Core::use_tx_gas(ctx, params.gas_costs.tx_undelegate)?;
+        let store_receipt = body.receipt > 0;
+        if store_receipt {
+            <C::Runtime as Runtime>::Core::use_tx_gas(ctx, params.gas_costs.store_receipt)?;
+        }
 
         // Check whether undelegate is allowed.
         if params.disable_undelegate {
             return Err(Error::Forbidden);
         }
+        // Make sure receipts can only be requested internally (e.g. via subcalls).
+        if store_receipt && !ctx.is_internal() {
+            return Err(Error::InvalidArgument);
+        }
 
         // Signer.
         let signer = &ctx.tx_auth_info().signer_info[0];
         let to = signer.address_spec.address();
-        let nonce = signer.nonce;
-        Self::undelegate(ctx, body.from, nonce, to, body.shares)
+        let nonce = if store_receipt {
+            body.receipt // Use receipt identifer as the nonce.
+        } else {
+            signer.nonce // Use signer nonce as the nonce.
+        };
+        Self::undelegate(ctx, body.from, nonce, to, body.shares, store_receipt)
```

Comment:
this is a fair amount of duplicated code. can you split out the common stuff into a function that takes the other functions as input and then does the verification and nonce setup?

Reply:
Duplication between delegate and undelegate you mean?